### PR TITLE
server: Add support for hostPorts

### LINKED ISSUE #560: Add support for host ports
We will need to add support for node ports in k8s using iptables. 
@rajatchopra Can you let us know the specifics of what we need to do for this?

### PR DESCRIPTION
**Pull request by apilloud**
This fixes #560, mostly. The common cases of connecting to the port via the host's IP both locally and remotely does work. Connecting locally by 'localhost:port' doesn't work, as that requires the interface which can reach the container to be known. It wasn't obvious to me on how to get that out of the netplugin.

@rajatchopra PTAL.

### CODE REVIEW

**Comment by mrunalp** [comment]
Can we add an integration test? Thanks!

**Comment by apilloud** [comment]
Yep, test coming soon. Need to get them running on my system first.

**Comment by rajatchopra** [comment]
>  It wasn't obvious to me on how to get that out of the netplugin.

It seems impossible without enforcing a new interface API. We will ultimately have to rely on the network plugin to do all the work anyway and this code will go away. Until then, we can examine the option of enforcing an API that will return the device name used by the netplugin (and in case there is none, then the localhost access will not work).

This work looks good at a glance. Will review again at next chance. Thanks.

**Comment by apilloud** [comment]
@rajatchopra @mrunalp This has tests now and is passing. Please take a look when you get a chance.

**Comment by mrunalp** [comment]
@apilloud Yeah, I will take a look. 
@runcom e2e tests haven't run on this one :/

**Comment by runcom** [comment]
retest

**Comment by mrunalp** [comment]
Tests still haven't returned on this one :/

**Comment by runcom** [comment]
kicked in CI directly now...

File: server/sandbox_stop.go [hunk 1]
```diff
@@ -35,6 +36,15 @@ func (s *Server) StopPodSandbox(ctx context.Context, req *pb.StopPodSandboxReque
 		return nil, err
 	}
 	if _, err := os.Stat(netnsPath); err == nil {
+		if err2 := s.hostportManager.Remove(sb.id, &hostport.PodPortMapping{
+			Name:         sb.name,
+			PortMappings: sb.portMappings,
+			HostNetwork:  false,
+		}); err2 != nil {
+			return nil, fmt.Errorf("failed to remove hostport for container %s in sandbox %s: %v",
```
Comment (mrunalp) [inline server/sandbox_stop.go]: Do you think it might be better to warn here rather than fail? (We warn when CNI teardown fails).

Comment (apilloud) [inline server/sandbox_stop.go]: I agree that warning would be better. I copied from TearDownPod below and didn't notice when it changed on rebase. (Though I remember resolving the merge conflict. Oops.) Updating now.

**Comment by apilloud** [comment]
This has been sitting around for a while, so I rebased.

**Comment by runcom** [comment]
Mmm something has failed in RHEL

**Comment by runcom** [comment]
The test is failing in rhel:

not ok 42 Connect to pod hostport from the host

**Comment by runcom** [comment]
More output from failing test:

```
 Ncat: You must specify a host to connect to. QUITTING.
RHEL
```

**Comment by apilloud** [comment]
A few lines up: `Error: either "dev" is duplicate, or "proto" is a garbage.` Something about looking up the default gateway failed. Added some debug output, as I don't have a RHEL system.

**Comment by mrunalp** [comment]
retest

**Comment by runcom** [comment]
Everything's OK here, LGTM

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**